src/Sylius/Bundle/SettingsBundle/Doctrine/EventListener/ParameterTransformerListener.php

```diff
@@ -0,0 +1,132 @@
+
+    /**
+     * @var array
+     */
+    protected $parametersMap = [];
+
+    public function __construct(ContainerInterface $container)
+    {
+        $this->container = $container;
+    }
+
+    /**
+     * @param LifecycleEventArgs $event
+     */
+    public function postLoad(LifecycleEventArgs $event)
+    {
+        $settings = $event->getObject();
+
+        if ($settings instanceof SettingsInterface) {
+            $this->reverseTransform($settings);
+        }
+    }
+
+    /**
+     * @param OnFlushEventArgs $event
+     */
+    public function onFlush(OnFlushEventArgs $event)
+    {
+        $em = $event->getEntityManager();
+        $uow = $em->getUnitOfWork();
+
+        foreach ($uow->getScheduledEntityInsertions() as $entity) {
+            if ($entity instanceof SettingsInterface) {
+                $this->transform($entity);
+            }
+        }
+
+        foreach ($uow->getScheduledEntityUpdates() as $entity) {
+            if ($entity instanceof SettingsInterface) {
+                $this->transform($entity);
+            }
+        }
+    }
+
+    public function postFlush()
+    {
+        // revert settings parameters to what they were before flushing
+        foreach ($this->parametersMap as $map) {
+            $map['entity']->setParameters($map['parameters']);
+        }
+
+        // reset parameters map
+        $this->parametersMap = [];
+    }
+
+    /**
+     * @param SettingsInterface $settings
+     */
+    protected function transform(SettingsInterface $settings)
+    {
+        // store old parameters, so we can revert to it after flush
+        $this->parametersMap[] = [
+            'entity' => $settings,
+            'parameters' => $settings->getParameters(),
+        ];
+
+        $transformers = $this->getTransformers($settings);
+        foreach ($settings->getParameters() as $name => $value) {
+            if (isset($transformers[$name])) {
+                $settings->set($name, $transformers[$name]->transform($value));
+            }
+        }
+    }
+
+    /**
+     * @param SettingsInterface $settings
+     */
+    protected function reverseTransform(SettingsInterface $settings)
+    {
+        $transformers = $this->getTransformers($settings);
+        foreach ($settings->getParameters() as $name => $value) {
+            if (isset($transformers[$name])) {
+                $settings->set($name, $transformers[$name]->reverseTransform($value));
+            }
+        }
+    }
+
+    /**
+     * @param SettingsInterface $settings
+     *
+     * @return ParameterTransformerInterface[]
+     */
+    protected function getTransformers(SettingsInterface $settings)
+    {
+        $registry = $this->container->get('sylius.settings.schema_registry');
+        $schema = $registry->get($settings->getSchemaAlias());
+
+        $settingsBuilder = new SettingsBuilder();
+        $schema->buildSettings($settingsBuilder);
+
+        return $settingsBuilder->getTransformers();
+    }
+}
```

Comment:
I have a feeling that this (not this specific implementation, my idea in general ;p) will hurt us in the long-run. I want to move default tax zone setting to Channel (see #4446), is there any other use-case for transformers at the moment? I think we could safely store codes of specific entities and do transformation/comparison whenever needed. I am really reluctant to add more doctrine listeners. Post load is called on every entity...